File: maximize-happiness-of-selected-children.ts
{
    function maximumHappinessSum(happiness: number[], k: number): number {
        happiness = happiness.sort((a,b) => a - b)
        let c = 0;

        for (let i = 0; i < k; i++) {
            const value = (happiness?.pop() || 1) - i;

            c += value < 0 ? 0 : value
        }
        

        return c;
    };

    console.log(maximumHappinessSum([1,2,3], 2))
}

File: maximize-happiness-of-selected-children.rs
use std::collections::BinaryHeap;

pub fn maximum_happiness_sum(happiness: Vec<i32>, k: i32) -> i64 {
    let mut pq: BinaryHeap<i32> = happiness.into_iter().collect();

    let mut res: i64 = 0;
    let mut neg = 0;
    for _ in 0..k {
        if let Some(top) = pq.pop() {
            let v = top - neg;
            neg += 1;
            if v <= 0 {
                break;
            }
            res += v as i64;
        } else {
            break;
        }
    }
    res
}
